### ISSUE TITLE
Set Coordinates' latitude and longitude

### ISSUE THREAD

**Issue by JulianBroudy** [sothawo/mapjfx#58]
Hello,

I have integrated your library into my JavaFX application.
Amazing API let me say. Easy to use, clean and elegant, so thank you.

I am wondering though, would it be possible to set the Coordinates' latitude and longitude dynamically rather than only by instantiation?

See, I would like to wrap the class containing a Coordinate attribute but it's not possible unless the attribute - the coordinates - has a getter and a setter. Seeing as Coordinate does not contain setter methods, as a workaround, I was going to extend your Coordinate class and provide the setter but unfortunately, the class is set to final.

Side note: the loading of the tiles seems to be VERY slow at times and others work excellent for some regions on the map. I only set the map's zoom and center and initialize it. I am using the default map type. 

It would be great if you could point me in the right direction.
Thanks in advance.
Julian.

**Comment by sothawo** [comment]
Hi,

glad to hear that mapjfx for you is useful.

The `Coordinate` is a final immutable class by purpose. What would it mean if it were mutable? If you take a marker for example, this now monitors it's position by having it wrapped in a `SimpleObjectProperty<Coordinate>`, so that when a new coordinate is set into the property, the marker will adjust it's position on the map. If the `Coordinate` were mutable, it would be necessary to monitor both the latitude and the longitude value of the coordinate, because a change in that values would need to be reflected in the map as well. So for mapjfx a coordinate is a fixed point that does not change.

Perhaps you can describe the problem you want to solve by making the `Coordinate` mutable in more deftail, so we might find a solution for that.

Concerning loading times: There is a issue for that (#29) this came up when OpenStrretmap switched the tile's URLs to https, but I honestly don't know why this is happening. It's something deep in the bowels of the WebView browser from JavaFX, but that is native code and not debuggable. I know it's a PITA, sometimes it's ok, and then it takes forever. For my tests I use either Bing Maps or the XYZ source from the test app / demo.

**Comment by JulianBroudy** [comment]
Well, I have a `MapsManager` class which contains:

1. An `ObjectProperty<MapDTO>` **bean**.
1. A `BeanPathAdapter<MapDTO>` **superBean**.
1. A `List<MapDTO>` **mapsList**.

 [BeanPathAdapter](https://github.com/JFXtras/jfxtras-labs/blob/8.0/src/main/java/jfxtras/labs/scene/control/BeanPathAdapter.java) is a class I got from [JFXTras Library](https://github.com/JFXtras). It makes bidirectional binding to/from the wrapped object's attributes possible without having to make them JavaFX properties. All that there is to know about it is that it requires the attribute to have a getter and a setter adhering to the Java naming conventions, i.e., "`newAttribute`, `getNewAttribute()`, `setNewAttribute()`".

The `MapsManager` **bean** and **superBean** are synchronized, i.e., their wrapped object is always the same.
In a `FXMLController` class I have a `ComboBox` that is bound to the `MapsManager`'s **mapsList**.
I also bind bidirectionally the `ComboBox`'s `selectedProperty` to the `MapsManager`'s **bean**.
Now in my `FXMLController` I would like to be able to do something like:
```
MapsManager.getInstance().getSuperBean()
        .bindBidirectional("coordinates", mapView.centerProperty(), Coordinate.class);
```

The same way I am able to do:
```
MapsManager.getInstance().getSuperBean()
                .bindBidirectional("zoom", mapView.zoomProperty());
```

Now I tried creating a wrapper class that would encapsulate `Coordinate` and provide a getter and a setter but it did not pan out. I tried a couple more options but 

> all roads lead to Rome.

There is no workaround for not having the setter. I even tried binding it to my wrapper class which provides decorative methods to manually set the Coordinate attribute but then I wasn't able to bind it to `mapView`'s `centerProperty`. I ended up overcomplicating it so I just used a listener on the **bean** and whenever it or its contents change I manually set the `mapView`'s `centerProperty`.


I was thinking Coordinate class could maybe encapsulate itself and provide a setter like this:
```
public class CoordinatesWrapper{

Coordinate coordinates = new Coordinate();

  public Coordinate getCoordinates() {
    return coordinates;
  }

  public void setCoordinates(Coordinate coordinates) {
    this.coordinates = coordinates;
  }

  public Double getLatitude() {
    return coordinates.getLatitude();
  }

  public Double getLongitude() {
    return coordinates.getLongitude();
  }
}
```

Side note: I did some checking and I am pretty sure that only the latitude or the longitude would be referred to as a "coordinate", while to both would be "coordinates". Please excuse my pedanticism.

If interested, would be more than happy to share my project, it is private right now.
Thank you, again.

**Comment by sothawo** [comment]
What do you want to achieve by `MapsManager.getInstance().getSuperBean()
        .bindBidirectional("coordinates", mapView.centerProperty(), Coordinate.class);`?

For observing a change in the `MapView`'s center (by panning the map with the mouse for example), the coordinate (I stick to the naming here) wouldn't need to be changed. This change is normally observable.

If you want to change the latitude/longitude values of the map's center without creating a new one, having the setters wouldn't help you to get the changed value into a `MapView` object, because `MapView` observes if a new object is passed as center, not if lat/lon of the existing object are changing. To have this feature, the Coordinate object itself not only would need setters, but also it would be necessary to change these values from `Double` to `SimpleDoubleProperty`.

Or dou you really just need a version of `Coordinate` that has a noarg ctor and setter method so that  _JFXtra_ is able to create a `Coordinate` object? Please fork this repo, create the setters and noarg ctor for the `Coordinate` class and see if this works for you. If yes, add this functionality and add an IllegalStateException if the setter tries to set a value to an already non-null lat/lon property, so that the object cannot be modified after the creation and then create a Pull Request. I'd accept this as a different way for creating `Coordinate` objects, but not for modifiying existing ones, because the ability to modify existing lat/lon values would mean that the map would have to reflect this change.

**Comment by JulianBroudy** [comment]
My main aim from the binding is changing the **beans**'s coordinates according to the `mapView`'s center Coordinates.
The functionality I am looking for is this:
choosing a map from the `mapsList` recenters the `mapView`'s center (because the **bean** has coordinates) but also when moving the map the **bean**'s center coordinates change along with it.
I will do what you said, thank you.

I am wondering, is there any reason the tiles won't load at all? meaning running the application on one computer works wonderfully but on another doesn't show any tile ever?
Thanks again.

**Comment by sothawo** [comment]
Alas I have no clue hy this is happening, see #29 for what I have found about this problem

**Comment by JulianBroudy** [comment]
I have seen this issue, it isn't quite the same.
In my case it doesn't load at all, how do I know?
whenever a map loads a rectangle appears in the bottom right corner, usually a small [i] box.
with OSM it loads rectangle with "OpenStreetMaps". 
What I am referring to doesn't get to this point. It is not that it's stuck on loading tiles per se, I think it gets stuck in the first phase of initializing the map view for some reason, something is blocking it.

Anyway, your replies are very much appreciated, thank you for your hard work.
If you don't have anything to add I will do what you said, fork and let you know if I get anywhere worth mentioning.

**Comment by sothawo** [comment]
can you increase the log level to trace and see if initialization is stuck somewhere?

**Comment by JulianBroudy** [comment]
I'm sorry it has been so long, I have a couple of prior commitments I must deal with before I can get back to this issue. But I sure will, as soon as I am done with them, check the log level as well as fork the repo. I am not sure though if I am supposed to close this issue for now?

**Comment by sothawo** [comment]
no, leave it open if you want to create a PR like I suggested [in this comment](https://github.com/sothawo/mapjfx/issues/58#issuecomment-506475716)

**Comment by JulianBroudy** [comment]
> can you increase the log level to trace and see if initialization is stuck somewhere?

Hello again, 
Do you mean to see if the initialization is stuck on my end?

**Comment by sothawo** [comment]
yes, I'd like to know if there arte some errors, exceptions or whatever is happening, and increasing the log level might give a hint, what was the last operation that was executed successfully

**Comment by JulianBroudy** [comment]
I am not able to replicate the error.
The current state is:

- Running perfectly on Java 8 (1.8.0_212)
- Shows only 1 tile in the top left corner as mentioned in issue #36. I can't seem to fix it. 
As a workaround, I am migrating my whole project to:
- Java 11 (Java 11.0.2 & JavaFX 11.0.2), getting exceptions on running. 

So I created an SSCCE to try & run MapJFX 2.8.0 but getting this exception:

`Exception in thread "JavaFX Application Thread" java.lang.RuntimeException: java.lang.reflect.InvocationTargetException`

This happens only when I add the MapView to the FXML file. 
Any idea? I can create & share the mini-project if it helps.

**Comment by sothawo** [comment]
can you check out the [mapjfx-demo](https://github.com/sothawo/mapjfx-demo) and run this? this works with Java 11 when building the 2.x or master branch. Otherwise if you can provide a sample project I can have a look at this.

**Comment by JulianBroudy** [comment]
Weirdly enough, mapjfx-demo runs without exceptions but the map loads the same way I [described](https://github.com/sothawo/mapjfx/issues/58#issuecomment-526570258) regarding Java 10 :

![image](https://user-images.githubusercontent.com/46682375/64035190-d43f7580-cb58-11e9-8423-9e034b838afe.png)

**Comment by JulianBroudy** [comment]
I uploaded a sample project and invited you as a collaborator.
This is the project in which I get an exception in Java 11.

**Comment by sothawo** [comment]
check the fixes.adoc file in your sample repo and the changes I made

**Comment by sothawo** [comment]
when the tiles are only loaded partially, this is the known problem where deep in the native code the threads hang

**Comment by JulianBroudy** [comment]
Thank you for your continuous support.
Now it does run but unfortunately isn't very useful because of the top left corner tile issue.
So as I gather, my only option is to go back to Java 8?

**Comment by sothawo** [comment]
the problem of not loading the tiles is not dependent on Java 8 or 11, we have this since October 2017 (see #29). How is it if you switch to using the sample xyz/wms servers that are set up in the mapjfx-demo project? 
Do you have the caching enabled for MapView?

**Comment by JulianBroudy** [comment]
> the problem of not loading the tiles is not dependent on Java 8 or 11, we have this since October 2017 (see #29).

For some reason, this does not sound right to me. I don't have caching enabled. Maybe I have no idea what I'm talking about here but to me, it seems caching isn't the issue here because on Java 8 it works perfectly. I also don't think this is an issue of **tiles not loading** (being stuck or slow loading) since I have the exact same project, running on Java 8, Java 10 & now - thanks to your recent help - Java 11.

**Java 8:**
![image](https://user-images.githubusercontent.com/46682375/64065407-f519d080-cc15-11e9-8f4b-95b0a2026e81.png)


**Java 10/11:**
![image](https://user-images.githubusercontent.com/46682375/64065411-05ca4680-cc16-11e9-8be2-40a1a6868d42.png)

Currently, I am using BingMaps. But I have also tried initializing the map with the xyz option directly, the same result. I now have this option enabled by toggling the "Aerial View":
 `private XYZParam xyzParams = new XYZParam().withUrl("https://server.arcgisonline.com/ArcGIS/rest/services/World_Topo_Map/MapServer/tile/{z}/{y}/{x})").withAttributions("'Tiles &copy; <a href=\"https://services.arcgisonline.com/ArcGIS/rest/services/World_Topo_Map/MapServer\">ArcGIS</a>'");` and switching between them gives the exact same end result.

To summarize:
1. Nowhere have I encountered a slow loading issue in my all recent experiments.
1. The whole mapView (all tiles when running on Java 8) **as well as** the 1 top left "tile" (Java 10+), load almost instantaneously. 
3. As you can see almost the whole view (coordinates + zoom are the same in the above pictures) is rendered. It doesn't make sense that some tiles aren't loaded, more like the whole mapView is in fact the small top-left rectangle rather than the "big" rectangle.
2. When running on Java 10/11, mapView's modifications like zooming in-&-out are also rendered almost instantaneously. The effect looks more like that the map's actual bounds aren't what I set them to be in the `FXML` file or Scene Builder. Almost like the mapView has been "shrunk" (resized) and moved a bit.

**Comment by sothawo** [comment]
ok, this is strange indeed. After seeing your screenshots I would have supposed that there might be some problem with the custom overlay you have for the [i] and the switch to aerial view, but then the mapjx-demo should have rendered correctly.

I played around with the Java 1 version myself and saw the occasionally I get errors logged out to stderr like

> javax.net.ssl.SSLException: No PSK available. Unable to resume.

Not knowing if this might be related, but searching I found [this SO question](https://stackoverflow.com/questions/52574050/javax-net-ssl-sslexception-no-psk-available-unable-to-resume) although disabling TLS  1.3 is no solution, as then the mpas don't load at all, seems to be a requirement from the server side.

How do you start the application? From a terminal a an IDE? Do you have access to the stderr out output to see if you have these errors as well? If I get them, parts of the map stay empty, as the loading was interrupted.

**Comment by JulianBroudy** [comment]
> ok, this is strange indeed. After seeing your screenshots I would have supposed that there might be some problem with the custom overlay you have for the [i] and the switch to aerial view, but then the mapjx-demo should have rendered correctly.

Well yeah me too, Also I already ran a version that initializes the map with only the XYZ parameters.

> How do you start the application? From a terminal a an IDE? Do you have access to the stderr out output to see if you have these errors as well? If I get them, parts of the map stay empty, as the loading was interrupted.

I use the IDE's built-in run option.
IntelliJ Ultimate.
If by stderr out output you mean the console? then yes I see it and no errors.

I'm inclined to do one of the following depending on **which would you prefer**:

1. Forget about this, continue working with Java 8.
2. Create 3 simple projects with only 1 FXML and a MapView, each running different Java version and invite you - assuming I'll keep getting same results as mentioned before.
3. Share my current project.

**Comment by sothawo** [comment]
I am using IJ Ultimate as well and have the errors in the console.

If you can have two branches for your project using java 8 and the other a higher version that might be a way to go, although it is more work to keep them updated; that's what I am doing with mapjfx.

I will create a branch based on 2.x and switch that to Java 12, you could either clone and build it then or I could put the jar file in a bin directory into the repository for downloading.

**Comment by JulianBroudy** [comment]
I'm sorry I'm not sure I'm following.
Why create a branch for Java 12?

> If you can have two branches for your project using java 8 and the other a higher version that might be a way to go, although it is more work to keep them updated; that's what I am doing with mapjfx.

That's what I'm doing now.

I could share that to test it on your machine & see if the same problem occurs. Then we know my computer is missing something..

**Comment by sothawo** [comment]
https://github.com/sothawo/mapjfx/blob/master/Downloads/mapjfx-2.8.0-12.jar
This is built using OpenJDK 12.0.2

btw on which system do you have this running?

**Comment by JulianBroudy** [comment]
Windows 10 Build 1903.

Do you mean I should migrate my project to Java 12 and see if the map loads fully?

**Comment by sothawo** [comment]
yes , might be worth trying

**Comment by JulianBroudy** [comment]
> Or dou you really just need a version of `Coordinate` that has a noarg ctor and setter method so that _JFXtra_ is able to create a `Coordinate` object?

So my objective is to be able to change existing Coordinate's longitude & latitude regardless of whether it is set as the MapView's center Coordinate or not. In other words, to be able to bind the Coordinate's longitude & latitude to some other two doubles.

Right now it is only possible by observing the other two doubles and creating a new Coordinate with each change.

**Comment by sothawo** [comment]
regarding a Coordinate object only it might be feasable to make the properties changeable, which would mean, that mapjfx would have to react to property changes and then adapt the object within the map. But this would be a breaking change for the users, as it would remove the immutability from the Coordinate object.

What would this mean for a CoordinateLine object? A CoordinateLine on creation copies the array of Coordinates so that the array cannot be mutated anymore. If the coordinates would change, then a Coordinate would have to monitor all it's Coordinates and would need to recreate the underlying OpenLayers polygon. Markers and Labels as well need to be adapted to not only react to the setCoordinate() method, but also to changes to the coordinates properties. 

It might be possible to introduce a new Object like ModifiableCoordinate in addition to the existing one which could be used for a Map's center or a Marker or Label's position, but I don't really see the benefit of adding this special functionality to the library. Such changes of a modifiable Coordinate internally would do nothing than what now can be done by setting a new Coordinate object as a position.

The same for the Marker's image url. If this needs to be changed, just create a new Marker object with the new image url at the same Coordinate then the old, then delete the old on and make the new visible.

**Comment by JulianBroudy** [comment]
Ok I understand, I was thinking to introduce anew kind of Coordinate as you say. 
But it seems like too much trouble for what I needed.

Two more quick questions before I close this issue as I won't be pursuing it:
1. Is it possible to set the map's center but also get the effect of mouse panning & zooming?
Right now there is no effect, it is like "loading" a different image. I would like - if possible - to get a transition effect between <code>current center</code>, <code>curent zoom</code> and <code>new center</code>, <code>new zoom</code>.

2. Would it be possible to get an option to lock map? as in disabling panning/zooming. 
Even better if there is an option to restrict panning/zooming to inside given boundaries. Meaning moving and zooming would be possible as long as mapView does not get out of pre-defined min/max top-left/buttom-right corners.

**Comment by sothawo** [comment]
1. do you mean an animation when setting the center or the zoom from within the program? Available since 17.12.2014 with `mapView.setAnimationDuration(500)` for example.

2. setting the `Configuration` option `interactive` to false locks the map. Or what do you mean with lock? As for the restriction, this could be implemented by setting a _bounding extent_ but there are open questions: what if let's say the map is zoomed out to the maximum allowed extend, and then the window size is enlarged. At the moment enlarging the window increases the visible area, for a bounded extent would this mean to zoom in? Can you please create a new issue for this where this can be discussed?
